pkg/arvo/sur/btc-node-hook.hoon

```diff
@@ -0,0 +1,1566 @@
+      $:  txid=@ux
+          vout=@ud
+          script-pubkey=@ux
+          redeem-script=(unit @ux)
+          witness-script=(unit @ux)
+          amount=@t
+      ==
+    ::  $raw-transaction-rpc-out:
+    ::
+    ::    Used in:
+    ::      - %get-block
+    ::      - %get-raw-transaction
+    ::
+    +$  raw-transaction-rpc-out
+      $:  in-active-chain=(unit ?)
+          hex=@ux
+          txid=@ux
+          hash=@ux
+          size=@ud
+          vsize=@ud
+          weight=@ud
+          version=@t
+          locktime=@ud
+          vin=(list vin)
+          vout=(list vout)
+          blockhash=(unit @ux)
+          confirmations=(unit @ud)
+          blocktime=(unit @ud)
+          time=(unit @ud)
+      ==
+    ::  $tx-in-block:
+    ::
+    ::    Used in:
+    ::      - %lists-in-ceblock
+    ::      - %list-transactions
+    ::
+    +$  tx-in-block
+      $:  address=(unit address)
+          =category
+          amount=@t
+          label=(unit @t)
+          vout=@ud
+          fee=(unit @t)
+          confirmations=@ud
+          blockhash=(unit @ux)
+          blockindex=(unit @ud)
+          blocktime=(unit @ud)
+          txid=@ux
+          time=@ud
+          time-received=(unit @ud)
+          wallet-conflicts=(unit (list @ux))
+          =bip125-replaceable
+          abandoned=(unit ?)
+          comment=(unit @t)
+          to=(unit @t)
+      ==
+    ::  $errors:
+    ::
+    ::    Used in:
+    ::      - %sign-raw-transaction-with-key
+    ::      - %sign-raw-transaction-with-wallet
+    ::
+    +$  errors  %-  list
+      $:  txid=@ux
+          vout=@ud
+          script-sig=@ux
+          sequence=@ud
+          error=@t
+      ==
+    ::  $mem-pool:
+    ::
+    ::    Used in:
+    ::      - $mem-pool-response
+    ::      - %get-mempool-entry
+    ::
+    +$  mem-pool
+      $:  size=@ud
+          fee=@t
+          modified-fee=@t
+          time=@ud
+          height=@ud
+          descendant-count=@ud
+          descendant-size=@ud
+          descendant-fees=@t
+          ancestor-count=@ud
+          ancestor-size=@ud
+          ancestor-fees=@t
+          w-txid=@ux
+          $=  fees
+          $:  base=@t
+              modified=@t
+              ancestor=@t
+              descendant=@t
+          ==
+          depends=(list @ux)
+          spent-by=(list @ux)
+          bip125-replaceable=?
+      ==
+    ::  $mem-pool-response:
+    ::
+    ::    Used in:
+    ::      - %get-raw-mempool
+    ::      - %get-mempool-entry
+    ::      -  %get-mempool-ancestors
+    ::      -  %get-mempool-descendants
+    ::      -  %get-raw-mempool
+    ::
+    +$  mem-pool-response  %-  list
+      ::  FIXME: a list for ux and map for the cell
+      ::  produces a fish-loop
+      ::
+      $?  ::  (for verbose = false)
+          ::
+          @ux                      :: (list @ux)
+          ::  (for verbose = true)
+          ::
+          [id=@ux =mem-pool]       :: (map id=@ux mem-pool)
+      ==
```

Comment:
I've moved all the `%-  unit` `%-  list` to the lines bellow.
I think you are right, at least when I experimented using lists and maps, all my attempts at making it work were in vane. 

Let me see if I follow your suggestion. The response is either a json array `(list @ux)`, or a json object `(list [@ux *])`, so it seems to me that the type will need to be a union, unless that, in the case of an array, you always produce `*mem-pool` for the tail of the pair?

Reply:
Are they complex types? imo this should be a `$@` not a `$?`, which has no semantic difference but hammers home "fork between atom and cell".

Reply:
> it seems to me that the type will need to be a union

Yes, but the type system is tripping up over that. So instead, I'm suggesting using `$%`, something like:

```hoon
+$  mem-pool-response
  $%  [%soft ids=(list @ux)]
      [%loud pool=(map @ux mem-pool)]
  ==
```

My additional suggestion was to provide a library function that `$-(mem-pool-response (list @ux))`, so that if you just want the `(list @ux)` from the response, you don't have to re-implement the destructuring every time. This may or may not make sense given the context, apply your best judgement.

Reply:
Ah right, my confusion came from me thinking that you were referring to the "current" one (list of either atom or cell). That doesn't seem to have issues with the type system?

I might end up using @ohAitch' (`$@`instead of `$?`), but leaving it as a list for both (I'll need to do a bit of rewriting to make it into a map). Thanks for the suggestions!